### PR TITLE
DirConstantFixer, ModernizeTypesCastingFixer, RandomApiMigrationFixer: [WIP]

### PR DESCRIPTION
**Pull request by kalessil**
Relevant fixers status:

- IsNull: not included (still my FB only)
- AliasFunctionsFixer: needs to be re-created on top of AbstractFunctionReferenceFixer
- 1.12 base PR not possible (UTs crashed, interfaces changes?)

Changes:

- AbstractFunctionReferenceFixer interfaces were consolidated with existing
- Block constructs detection re-factored according hints
- Extended UTs cases

Still open:

- Risky/nor risky notes for all 3 fixers (@keradus : feedback needed)

### CODE REVIEW

**Comment by kalessil** [comment]
@keradus @SpacePossum : finally done

File: README.rst [hunk 1]
```diff
@@ -541,6 +554,11 @@ Choose from the list of available fixers:
                         the class name should match
                         the file name. (Risky fixer!)
 
+* **random_api_migration**
```
Comment (GrahamCampbell) [inline README.rst]: What about replacing with `random_int`. That's the best function to use.

Comment (kalessil) [inline README.rst]: That's PHP7 only, mt_* are existing since PHP 4 so I have to stick to them.

Comment (GrahamCampbell) [inline README.rst]: I still think we need an option to enable fixing to that. Especially since the `random_compat` package exists for PHP 5.

Comment (GrahamCampbell) [inline README.rst]: I use `random_int` over `mt_rand` in my PHP 5 apps.

Comment (kalessil) [inline README.rst]: Is there any language level settings in CS fixer?

Comment (GrahamCampbell) [inline README.rst]: I don't think the php version matters. You may still want to fix to `random_int` even on PHP 5.

Comment (GrahamCampbell) [inline README.rst]: Infact, I'm telling you I do, lol.

Comment (kalessil) [inline README.rst]: It was late evening in Germany =)

For me it's not problem to support additional replacement options,  e.g. in .php_cs
```php
\Symfony\CS\Fixer\Contrib\RandomApiMigrationFixer::setReplacements(array(
            'rand' => '<whatever>',
            'srand' => '<whatever>',
            'getrandmax' => '<whatever>',
));
```

I can add this If this approach fits fixer's architecture. Any feedback on this?

Comment (SpacePossum) [inline README.rst]: it could be a config option on 2.x line, however I like the current setup :)
(since most of my projects are not yet on PHP 7 / `symfony/polyfill`)

**Comment by SpacePossum** [comment]
thanks so much @kalessil  :)
:+1:

File: README.rst (hunk 1 above)
Comment (keradus) [inline README.rst]: Please follow 2.x interface for it, eg: https://github.com/FriendsOfPHP/PHP-CS-Fixer/blob/master/Symfony/CS/Fixer/Contrib/PhpUnitConstructFixer.php#L39

(Just remember to make it static)

Comment (kalessil) [inline README.rst]: Thanks for the hint, will do.

**Comment by stof** [comment]
I think this should be done in 1.x versions though (the interface of the fixer is indeed a bit different, but it is easy to convert it to 1.x: just create the Tokens at the beginning based on the code, and return the new code at the end)

File: README.rst (hunk 1 above)
Comment (kalessil) [inline README.rst]: Added configuration method: https://github.com/kalessil/PHP-CS-Fixer/blob/features_dirconst_randapi_typecast/Symfony/CS/Fixer/Contrib/RandomApiMigrationFixer.php#L37

@keradus : configure method already defined in a parent, so making it static not possible

**Comment by kalessil** [comment]
@stof : would you please elaborate why (I tried and not sure if it worth investing time into adjustments)?

File: README.rst (hunk 1 above)
Comment (keradus) [inline README.rst]: I have been sure you are sending the PR to 1.12 branch, as stated in contributing file...
And on 1.12 it would be static

Comment (kalessil) [inline README.rst]: Okay, moving on then;

**Comment by keradus** [comment]
Everything would be fine if you follow contributing guide before opening PR.
OK, so what we can do now...

How many work is needed to backport this PR to 1.x line? If a lot - please point out the most problematic things. Then, it would be pointless to waste time to backport to 1.x and then, again, waste time to merge it to master. But everything depend on how problematic the backporting would be.

**Comment by GrahamCampbell** [comment]
Would be easier to just do this on master :trollface:

**Comment by kalessil** [comment]
Okay, research status: \Symfony\CS\Tokenizer\Token(s) in master are not back compatible with 1.x:

\Symfony\CS\Tokenizer\Tokens::overrideRange is not exists in 1.12
PHP Warnings: use of undefined constant CT_NAMESPACE_OPERATOR 

I can not back-port this, master are 1.x are not compatible.
On other side this PR is introducing only new files, so merging 1.x->master should be still possible.

**Comment by kalessil** [comment]
@GrahamCampbell: that's different, fixer's API in master allows to implement more complex features.

File: Symfony/CS/AbstractFunctionReferenceFixer.php [hunk 2]
```diff
@@ -0,0 +1,134 @@
...
+            if ($token->equals(',')) {
+                ++$argumentsCount;
+            }
+        }
+
+        return $argumentsCount;
+    }
+
+    /**
+     * Checks if function/method defined in the scope.
```
Comment (keradus) [inline Symfony/CS/AbstractFunctionReferenceFixer.php]: `is defined` ?

**Comment by keradus** [comment]
OK, I remember that already point it out in some old PR.

Do not try to guess if the function is defined somewhere. You may check only the current file, not any external includes. So this check can not work properly. So is is pointless to try to guess it. Remove that part.

What should be check is to check if function is called from global scope,
so if it is:
* `srand()`
* `\srand()`

and not:
* `Foo::srand()`
* `Foo\srand()`
* `$foo->srand()`

File: README.rst (hunk 1 above)
Comment (keradus) [inline README.rst]: so... `configure` ?

File: Symfony/CS/Fixer/Contrib/RandomApiMigrationFixer.php [hunk 3]
```diff
@@ -0,0 +1,98 @@
...
+        'getrandmax' => 'mt_getrandmax',
+    );
+
+    /**
+     * Static analog of 'public function configure(array $configuration = null)',
+     * which can not be overridden in favor of static.
+     *
+     * @param string[]|null $customReplacements
+     */
+    public static function configureReplacement(array $customReplacements = null)
```
Comment (keradus) [inline Symfony/CS/Fixer/Contrib/RandomApiMigrationFixer.php]: This is not tested.

File: Symfony/CS/Fixer/Contrib/RandomApiMigrationFixer.php [hunk 4]
```diff
@@ -0,0 +1,98 @@
...
+    {
+        if (null !== $customReplacements) {
+            foreach (self::$replacements as $pattern => &$replacement) {
+                if (array_key_exists($pattern, $customReplacements) && is_string($customReplacements[$pattern])) {
```
Comment (keradus) [inline Symfony/CS/Fixer/Contrib/RandomApiMigrationFixer.php]: if input data is not valid throw a proper exception

File: Symfony/CS/Tests/Fixer/Contrib/RandomApiMigrationFixerTest.php [hunk 5]
```diff
@@ -0,0 +1,77 @@
...
+    echo srand;
+}
+}
+
+class srand extends SrandClass{
+const srand = "srand"
+}
+', ),
+            array('<?php mt_srand($a);', '<?php srand($a);'),
+            array('<?php \\mt_srand($a);', '<?php \\srand($a);'),
```
Comment (keradus) [inline Symfony/CS/Tests/Fixer/Contrib/RandomApiMigrationFixerTest.php]: invalid code:
https://3v4l.org/Ypn36

Comment (GrahamCampbell) [inline Symfony/CS/Tests/Fixer/Contrib/RandomApiMigrationFixerTest.php]: isn't the first \ to escape the 2nd?

Comment (GrahamCampbell) [inline Symfony/CS/Tests/Fixer/Contrib/RandomApiMigrationFixerTest.php]: https://3v4l.org/eGEY8

Comment (keradus) [inline Symfony/CS/Tests/Fixer/Contrib/RandomApiMigrationFixerTest.php]: OK, then...
there is no need to escape it:
https://3v4l.org/4eBLq

**Comment by keradus** [comment]
BTW, please fix the build ;)

File: Symfony/CS/AbstractFunctionReferenceFixer.php [hunk 6]
```diff
@@ -0,0 +1,134 @@
...
+abstract class AbstractFunctionReferenceFixer extends AbstractFixer
+{
+    /**
+     * Looks up Tokens sequence for suitable candidates and delivers boundaries information,
+     * which can be supplied by other methods in this abstract class.
+     *
+     * @param string $functionNameToSearch
+     * @param Tokens $tokens
+     * @param int    $start
+     * @param int    $end
```
Comment (keradus) [inline Symfony/CS/AbstractFunctionReferenceFixer.php]: or null ?

**Comment by keradus** [comment]
ping @kalessil

**Comment by keradus** [comment]
ping @kalessil

**Comment by kalessil** [comment]
I'm here, should finish this evening

File: README.rst (hunk 1 above)
Comment (kalessil) [inline README.rst]: yep: `configure`, added UTs for it as well

**Comment by kalessil** [comment]
Merged throw annotations into one

**Comment by kalessil** [comment]
@keradus : fixing build. What exactly I need to add to fixers descriptions to mark them as risky?

Personally I'd prefer to have a method in fixers (isRisky()), so --no-risky run flag/consistent descriptions can be introduced.

**Comment by keradus** [comment]
> What exactly I need to add to fixers descriptions to mark them as risky?
> 
> Personally I'd prefer to have a method in fixers (isRisky()), so --no-risky run flag/consistent descriptions can be introduced.

[this](https://github.com/FriendsOfPHP/PHP-CS-Fixer/blob/master/Symfony/CS/FixerInterface.php#L55), [this](https://github.com/FriendsOfPHP/PHP-CS-Fixer/blob/master/Symfony/CS/Console/Command/FixCommand.php#L110) and [that](https://github.com/FriendsOfPHP/PHP-CS-Fixer/blob/master/Symfony/CS/Config/Config.php#L199)

**Comment by kalessil** [comment]
=) Cool, thanks!

**Comment by kalessil** [comment]
Okay, I worked out all remarks. All 3 fixers marked as risky.

**Comment by keradus** [comment]
@kalessil please, could you rebase the branch ?

**Comment by kalessil** [comment]
@keradus : rebased

**Comment by SpacePossum** [comment]
very nice :)  
:+1:

File: README.rst (hunk 1 above)
Comment (keradus) [inline README.rst]: You have test the method's setter behaviour, but not the change for fixer behaviour.
Check this out: https://github.com/FriendsOfPHP/PHP-CS-Fixer/blob/30acb92898d474e5caa4da5314d281e889889f9f/Symfony/CS/Tests/Fixer/Contrib/PhpUnitConstructFixerTest.php#L36

**Comment by kalessil** [comment]
@keradus : UTs for RandomApiMigrationFixer adjusted (take into account re-configured replacements)

File: src/Fixer/Contrib/DirConstantFixer.php [hunk 7]
```diff
@@ -0,0 +1,86 @@
...
+/*
+ * This file is part of the PHP CS utility.
+ *
+ * (c) Fabien Potencier <fabien@symfony.com>
+ *
+ * This source file is subject to the MIT license that is bundled
+ * with this source code in the file LICENSE.
+ */
+
+namespace Symfony\CS\Fixer\Contrib;
```
Comment (gharlan) [inline src/Fixer/Contrib/DirConstantFixer.php]: symfony level?

File: src/AbstractFunctionReferenceFixer.php [hunk 8]
```diff
@@ -0,0 +1,125 @@
...
+    protected function countArguments(Tokens $tokens, $openParenthesis, $closeParenthesis)
+    {
+        $firstSensibleToken = $tokens->getNextMeaningfulToken($openParenthesis);
+        if ($tokens[$firstSensibleToken]->equals(')')) {
+            return 0;
+        }
+
+        $argumentsCount = 1;
+        for ($paramContentIndex = $openParenthesis + 1; $paramContentIndex < $closeParenthesis; ++$paramContentIndex) {
+            /* @var \Symfony\CS\Tokenizer\Token $token */
```
Comment (gharlan) [inline src/AbstractFunctionReferenceFixer.php]: should not be necessary

File: src/Fixer/Contrib/ModernizeTypesCastingFixer.php [hunk 9]
```diff
@@ -0,0 +1,118 @@
...
+ * @author Vladimir Reznichenko <kalessil@gmail.com>
+ */
+final class ModernizeTypesCastingFixer extends AbstractFunctionReferenceFixer
+{
+    /**
+     * {@inheritdoc}
+     */
+    public function isCandidate(Tokens $tokens)
+    {
+        return $tokens->isTokenKindFound(T_STRING);
```
Comment (gharlan) [inline src/Fixer/Contrib/ModernizeTypesCastingFixer.php]: maybe check the casting tokens here?

File: src/Fixer/Contrib/DirConstantFixer.php (hunk 7 above)
Comment (gharlan) [inline src/Fixer/Contrib/DirConstantFixer.php]: hmm, it is risky. Do we use risky fixers in symfony level?

File: src/Fixer/Contrib/ModernizeTypesCastingFixer.php (hunk 9 above)
Comment (kalessil) [inline src/Fixer/Contrib/ModernizeTypesCastingFixer.php]: No need,  we want to replace function calls with type casting.

File: README.rst (hunk 1 above)
Comment (kalessil) [inline README.rst]: Was done already.

Comment (keradus) [inline README.rst]: I can't see any. You always check default configuration

File: src/Fixer/Contrib/RandomApiMigrationFixer.php [hunk 10]
```diff
@@ -0,0 +1,95 @@
...
+     */
+    public function configure(array $customReplacements = null)
+    {
+        if (null !== $customReplacements) {
+            foreach ($customReplacements as $pattern => $replacement) {
+                if (!array_key_exists($pattern, self::$replacements)) {
+                    throw new \UnexpectedValueException(sprintf('"%s" is not handled by the fixer', $pattern));
+                }
+                if (!is_string($replacement)) {
+                    throw new \InvalidArgumentException(sprintf('Expected string got "%s"', gettype($replacement)));
```
Comment (keradus) [inline src/Fixer/Contrib/RandomApiMigrationFixer.php]: `InvalidFixerConfigurationException` x2

File: src/Fixer.php [hunk 11]
```diff
@@ -254,9 +254,9 @@ public function fixFile(\SplFileInfo $file, array $fixers, $dryRun, $diff, FileC
             if (!$dryRun && false === @file_put_contents($file->getRealPath(), $new)) {
                 $error = error_get_last();
                 if ($error) {
-                    throw new IOException(sprintf('Failed to write file "%s", "%s".', $file->getRealpath(), $error['message']), 0, null, $file->getRealpath());
+                    throw new IOException(sprintf('Failed to write file "%s", "%s".', $file->getRealPath(), $error['message']), 0, null, $file->getRealPath());
                 }
-                throw new IOException(sprintf('Failed to write file "%s".', $file->getRealpath()), 0, null, $file->getRealpath());
+                throw new IOException(sprintf('Failed to write file "%s".', $file->getRealPath()), 0, null, $file->getRealPath());
```
Comment (keradus) [inline src/Fixer.php]: wasn't this done by another PR?

File: src/Fixer/Contrib/RandomApiMigrationFixer.php [hunk 12]
```diff
@@ -0,0 +1,95 @@
...
+    private static $replacements = array(
+        'rand' => 'mt_rand',
+        'srand' => 'mt_srand',
+        'getrandmax' => 'mt_getrandmax',
+    );
+
+    /**
+     * @param string[]|null $customReplacements
+     */
+    public function configure(array $customReplacements = null)
```
Comment (keradus) [inline src/Fixer/Contrib/RandomApiMigrationFixer.php]: how to turn off replacing for example `getrandmax` ?

File: src/Fixer/Contrib/DirConstantFixer.php (hunk 7 above)
Comment (keradus) [inline src/Fixer/Contrib/DirConstantFixer.php]: we do not

File: README.rst (hunk 1 above)
Comment (kalessil) [inline README.rst]: https://github.com/kalessil/PHP-CS-Fixer/blob/features_dirconst_randapi_typecast/tests/Fixer/Contrib/RandomApiMigrationFixerTest.php#L104

Uses PHP7 function from configuration test

File: src/Fixer.php (hunk 11 above)
Comment (kalessil) [inline src/Fixer.php]: It was. Perhaps merging went wrong or this was new code added in master.

File: README.rst (hunk 1 above)
Comment (keradus) [inline README.rst]: And then you do not test `rand` -> `mt_rand`, please check the ref with example I provided.

File: src/Fixer/Contrib/RandomApiMigrationFixer.php (hunk 12 above)
Comment (kalessil) [inline src/Fixer/Contrib/RandomApiMigrationFixer.php]: I can add turning replacements individually off (by setting up replacement to null). Shall I do this?

File: tests/Fixer/Contrib/RandomApiMigrationFixerTest.php [hunk 13]
```diff
@@ -0,0 +1,110 @@
...
+    {
+        $this->getFixer()->configure(array('rand' => 'random_int'));
+
+        /* @var $aliases string[] */
+        $replacements = static::getStaticAttribute('\Symfony\CS\Fixer\Contrib\RandomApiMigrationFixer', 'replacements');
+        static::assertSame($replacements['rand'], 'random_int');
+    }
+
+    /**
+     * @depends testConfigure
```
Comment (keradus) [inline tests/Fixer/Contrib/RandomApiMigrationFixerTest.php]: I don't like that...
It could be done without it.

Any thoughts on that?

File: src/Fixer/Contrib/RandomApiMigrationFixer.php (hunk 12 above)
Comment (keradus) [inline src/Fixer/Contrib/RandomApiMigrationFixer.php]: This is one option (let call it patch the configuration).
The other one is replace the configuration (so if no key in input - no key in output).

We should decide a way that configuration method should behave and follow that everywhere.
Need to investigate it a bit...

File: README.rst (hunk 1 above)
Comment (kalessil) [inline README.rst]: I did, and it different from my case. Default are tested here:

```php
            array('<?php a(mt_srand());', '<?php a(srand());'),
            array('<?php a(\\mt_srand());', '<?php a(\\srand());'),
```

File: tests/Fixer/Contrib/RandomApiMigrationFixerTest.php [hunk 14]
```diff
@@ -0,0 +1,110 @@
...
+    /**
+     * @expectedException \Symfony\CS\ConfigurationException\InvalidFixerConfigurationException
+     */
+    public function testConfigureCheckSearchFunction()
+    {
+        $this->getFixer()->configure(array('is_null' => 'random_int'));
+    }
+
+    /**
+     * @expectedException \Symfony\CS\ConfigurationException\InvalidFixerConfigurationException
```
Comment (keradus) [inline tests/Fixer/Contrib/RandomApiMigrationFixerTest.php]: ex msg should be tested as well

File: tests/Fixer/Contrib/RandomApiMigrationFixerTest.php [hunk 15]
```diff
@@ -0,0 +1,110 @@
...
+    public function testConfigureCheckReplacementType()
+    {
+        $this->getFixer()->configure(array('rand' => null));
+    }
+
+    public function testConfigure()
+    {
+        $this->getFixer()->configure(array('rand' => 'random_int'));
+
+        /* @var $aliases string[] */
```
Comment (keradus) [inline tests/Fixer/Contrib/RandomApiMigrationFixerTest.php]: `$aliases` ?

File: tests/Fixer/Contrib/RandomApiMigrationFixerTest.php (hunk 13 above)
Comment (kalessil) [inline tests/Fixer/Contrib/RandomApiMigrationFixerTest.php]: I need to play around with UTs

File: src/Fixer/Contrib/ModernizeTypesCastingFixer.php (hunk 9 above)
Comment (gharlan) [inline src/Fixer/Contrib/ModernizeTypesCastingFixer.php]: indeed, sorry ;)

File: src/Fixer/Contrib/RandomApiMigrationFixer.php (hunk 12 above)
Comment (kalessil) [inline src/Fixer/Contrib/RandomApiMigrationFixer.php]: Ok, ping me then (and mention final decision in contributing notes please)

File: tests/Fixer/Contrib/RandomApiMigrationFixerTest.php (hunk 14 above)
Comment (kalessil) [inline tests/Fixer/Contrib/RandomApiMigrationFixerTest.php]: Will add

File: tests/Fixer/Contrib/RandomApiMigrationFixerTest.php (hunk 15 above)
Comment (kalessil) [inline tests/Fixer/Contrib/RandomApiMigrationFixerTest.php]: Ooops, will fix

**Comment by keradus** [comment]
BTW @kalessil , what about first two points from `Relevant fixers status:` in first topic ?

**Comment by kalessil** [comment]
It was one more fixer 'is_null(...)' -> 'null === ...' which not included in the PR
and a Symfony level fixer (alias functions) needs to be re-factored.

**Comment by keradus** [comment]
Yep, exactly that. Any chances to get that here as well ?
alias should be in this PR to really have a common interface and prove that base class is generic

**Comment by GrahamCampbell** [comment]
> It was one more fixer 'is_null(...)' -> 'null === ...' which not included in the PR

Is there any chance we could have an option to have such fixers not use yoda conditions?

**Comment by kalessil** [comment]
For me it's ok. Then I can propose following steps:

* get thins finished (you have to define common configuration requirements. me to full-fill them)
* refactor alias functions fixer (with configuration support?), when finished next step

As separate PR (so people could start using approved fixers):
* include is_null (if I understood right)
* finish uniqid fixer

**Comment by kalessil** [comment]
@GrahamCampbell : will provide configuration for this

**Comment by GrahamCampbell** [comment]
> @GrahamCampbell : will provide configuration for this

:+1: sexy

File: src/AbstractFunctionReferenceFixer.php [hunk 16]
```diff
@@ -0,0 +1,124 @@
...
+            if ($namespaceCandidateToken->isGivenKind(array(T_NEW, T_STRING, CT_NAMESPACE_OPERATOR))) {
+                // here can be added complete namespace scan
+                // this expression is differs from expected, resume
+                return $this->find($functionNameToSearch, $tokens, $openParenthesis, $end);
+            }
+        }
+
+        // final step: find closing parenthesis
+        $closeParenthesis = $tokens->findBlockEnd(Tokens::BLOCK_TYPE_PARENTHESIS_BRACE, $openParenthesis);
+        if (null === $closeParenthesis) {
```
Comment (SpacePossum) [inline src/AbstractFunctionReferenceFixer.php]: missing unit test for this case (when `true`)

File: src/AbstractFunctionReferenceFixer.php [hunk 17]
```diff
@@ -0,0 +1,124 @@
...
+     * @param int    $openParenthesis
+     * @param int    $closeParenthesis
+     *
+     * @return int
+     */
+    protected function countArguments(Tokens $tokens, $openParenthesis, $closeParenthesis)
+    {
+        $firstSensibleToken = $tokens->getNextMeaningfulToken($openParenthesis);
+        if ($tokens[$firstSensibleToken]->equals(')')) {
+            return 0;
```
Comment (SpacePossum) [inline src/AbstractFunctionReferenceFixer.php]: missing test for this case

File: src/Fixer/Contrib/ModernizeTypesCastingFixer.php (hunk 9 above)
Comment (SpacePossum) [inline src/Fixer/Contrib/ModernizeTypesCastingFixer.php]: maybe `T_FUNCTION` then?

Comment (gharlan) [inline src/Fixer/Contrib/ModernizeTypesCastingFixer.php]: `intval(1);` -> there is no `T_FUNCTION` token.. ;)

Comment (kalessil) [inline src/Fixer/Contrib/ModernizeTypesCastingFixer.php]: Can not find any reference for the constant, is it generated for function keyword or for 'call()'?

Comment (SpacePossum) [inline src/Fixer/Contrib/ModernizeTypesCastingFixer.php]: yeah, I wasn't thinking straight here ;)

**Comment by kalessil** [comment]
ping @keradus : should I finish this or wait for feedback regarding configuration requirements?

**Comment by keradus** [comment]
I won't be able to check this before Monday :(

**Comment by kalessil** [comment]
no problems, i'll working out open remarks then.

File: src/Fixer/Contrib/RandomApiMigrationFixer.php (hunk 12 above)
Comment (keradus) [inline src/Fixer/Contrib/RandomApiMigrationFixer.php]: Only `PhpUnitConstructFixer` allow to patch configuration, all other fixers only allow to replace configuration, and I would go in that way (for that PhpUnit as well, but maybe in separate PR)

**Comment by keradus** [comment]
[here](https://github.com/FriendsOfPHP/PHP-CS-Fixer/pull/1617#discussion_r49540026)

**Comment by keradus** [comment]
ping @kalessil

**Comment by kalessil** [comment]
Pretty clear: complete replacement of configuration, not partial testing.

PS: will rebase as well, regarding timeline hopefully have time tomorrow

**Comment by keradus** [comment]
ping @kalessil ;) I would really love to have it merged !

**Comment by keradus** [comment]
ping @kalessil

**Comment by keradus** [comment]
ping @kalessil

**Comment by keradus** [comment]
@kalessil ?

**Comment by kalessil** [comment]
@keradus : no progress so far, I'm almost offline until the end of February.

This PR has prio 1 in my open source activity list.

**Comment by keradus** [comment]
Thank you @kalessil for information.
Please, could you list down what left to be done? is the configuration thing the last one?

Maybe we could help finish this PR.

**Comment by kalessil** [comment]
Help would be really appreciated.

This PR:
- RandomApiMigrationFixer::configure needs to replace configuration (currently partially overriding)
- AbstractFunctionReferenceFixer : doesn't have UTs, covered partially by fixers test

It was proposal to include is_null and migrate AliasFunctionsFixer to AbstractFunctionReferenceFixer, but this should not happen in this PR IMO. So ideally just 2 items, which I delaying since mid of January.

File: src/AbstractFunctionReferenceFixer.php (hunk 17 above)
Comment (keradus) [inline src/AbstractFunctionReferenceFixer.php]: also, buggy for `<?php fnc($a, $b = array(), $c = 1) {}`

File: src/AbstractFunctionReferenceFixer.php (hunk 16 above)
Comment (keradus) [inline src/AbstractFunctionReferenceFixer.php]: in fact that case could only appear for invalid PHP code

**Comment by keradus** [comment]
@kalessil, please could you update the PR?

```bash
git remote add keradus git@github.com:keradus/PHP-CS-Fixer.git
git fetch keradus
git checkout 5da59cd7db72146bb44a3c7aaa7bab6ac99172e0
git push origin features_dirconst_randapi_typecast -f
```

ref https://github.com/keradus/PHP-CS-Fixer/commits/contrib-1617_replace

**Comment by keradus** [comment]
Very well. Replaced by #1819
